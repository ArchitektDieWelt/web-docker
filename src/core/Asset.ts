export type AssetPosition = "head" | "body";

export type AssetBase = {
  src: string;
  name?: string;
  position?: AssetPosition;
  priority?: number;
  countries?: string[];
};

export type AssetScript = AssetBase & {
  type: "js";
  async?: boolean;
  buildType?: "modern" | "legacy";
  deferred?: boolean;
};

export type AssetLink = AssetBase & {
  type: "css";
  media?: string;
};

export type Asset = AssetLink | AssetScript;
