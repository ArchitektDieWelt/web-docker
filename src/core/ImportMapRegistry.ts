interface ImportMap {
  [key: string]: string;
}

export class ImportMapRegistry {
  private importMap: ImportMap = {};

  constructor(
    private readonly script: HTMLScriptElement = document.createElement("script"),
    private readonly documentHead = document.head
  ) {
    this.script.type = "importmap";
    this.documentHead.appendChild(this.script);
  }

  public add(importMap: ImportMap) {
    this.importMap = {
      ...this.importMap,
      ...importMap,
    };
    this.script.textContent = JSON.stringify(this.importMap);
  }

  public get(): ImportMap {
    return this.importMap;
  }
}
