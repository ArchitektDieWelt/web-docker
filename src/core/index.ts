import RegisterEvent, { RegisterEventType } from "~/core/RegisterEvent";
import InitializeEvent, { InitializeEventType } from "~/core/InitializeEvent";

export { default as initialize } from "./initialize";

declare global {
  interface WindowEventMap {
    [RegisterEventType]: RegisterEvent;
    [InitializeEventType]: InitializeEvent;
  }
  interface Window {
    openreply: {
      initialize: (config: {
        configFilePath: string;
      }) => void;
    };
  }
}
