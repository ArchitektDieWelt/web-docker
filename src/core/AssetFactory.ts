import { Asset, AssetLink, AssetPosition, AssetScript } from "~/core/Asset";

export default class AssetFactory {
  private static buildTagScript(attr: AssetScript): HTMLScriptElement {
    const script = document.createElement("script");
    script.src = attr.src;
    if (attr.buildType === "modern") {
      script.type = "module";
      script.setAttribute("crossorigin", "");
    } else {
      script.type = "text/javascript";
      if (attr.async) script.setAttribute("async", "");
    }

    if (attr.priority)
      script.setAttribute("data-priority", attr.priority.toString());

    return script;
  }

  private static buildTagLink(attr: AssetLink): HTMLLinkElement {
    const link = document.createElement("link");
    link.setAttribute("rel", "stylesheet");
    link.setAttribute("href", attr.src);
    if (attr.media) link.setAttribute("media", attr.media);

    if (attr.priority)
      link.setAttribute("data-priority", attr.priority.toString());

    return link;
  }

  public create(
    assets: Asset[],
    position: AssetPosition = "head",
    countryCode?: string
  ): (HTMLLinkElement | HTMLScriptElement)[] {
    return (
      countryCode
        ? assets.filter(
            (asset) =>
              typeof asset.countries !== "object" ||
              asset.countries.includes(countryCode)
          )
        : assets
    )
      .filter((asset) => (asset.position ?? "head") === position)
      .sort((a, b) => {
        if (!a.priority || !b.priority) return 0;
        if (a.priority < b.priority) {
          return 1;
        } else if (a.priority > b.priority) {
          return -1;
        }
        return 0;
      })
      .map((entry) => {
        switch (entry.type) {
          case "css":
            return AssetFactory.buildTagLink(entry);
          case "js":
            return AssetFactory.buildTagScript(entry);
        }
      })
      .filter((item): item is HTMLLinkElement | HTMLScriptElement => !!item);
  }
}
