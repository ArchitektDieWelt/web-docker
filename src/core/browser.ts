import { InitializeEventType } from "~/core/InitializeEvent";
import initialize, { WebDockerOptions } from "~/core/initialize";

window.webdocker = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<WebDockerOptions>) => {
    await initialize(event.detail);
  }
);

