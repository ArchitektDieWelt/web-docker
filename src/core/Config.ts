import { ModuleConfig } from "~/core/ModuleConfig";

export type Config = ModuleConfig;
