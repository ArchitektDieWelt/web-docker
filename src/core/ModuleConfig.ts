import { Asset } from "~/core/Asset";
import { Config } from "~/core/Config";

export type IncludeType = "observed" | "page";

export interface ModuleConfigBase {
  version: string;
  assets: Asset[];
  module: string;
  type: IncludeType;
}

export interface ObservedModuleConfig extends ModuleConfigBase {
  type: "observed";
  selector: string;
}

export interface PageInclude {
  include: string;
  exclude?: string;
  exact?: boolean;
}

export interface PageModuleConfig extends ModuleConfigBase {
  type: "page";
  pages: (string | PageInclude)[];
}

export type ModuleConfig = ObservedModuleConfig | PageModuleConfig;

export class ModuleConfigService {
  public getModuleConfig(config: Config): ModuleConfig {
    if (!config.version)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the version field`
      );

    if (!config.assets)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the assets field`
      );

    if (!config.module)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the module field`
      );

    if (!config.type)
      throw Error(
        `Your tried to register a config with new structure: ${JSON.stringify(
          config
        )}  it was missing the type field`
      );

    if (config.type === "page" && !config.pages)
      throw Error(
        `Your tried to register a config with new structure of page include type:  ${JSON.stringify(
          config
        )}  it was missing the pages field`
      );

    return config;
  }
}
