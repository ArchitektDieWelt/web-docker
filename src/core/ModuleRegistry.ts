import { ModuleService } from "./ModuleService";
import AssetFactory from "~/core/AssetFactory";
import { Logger } from "~/core/Logger";
import { Config } from "~/core/Config";
import { ModuleConfig } from "~/core/ModuleConfig";
import { PageModuleService } from "~/core/PageModuleService";
import { ObservedModuleService } from "~/core/ObservedModuleService";

export interface ModuleRegistryInterface {
  add(config: Config): ModuleService;
}

class ModuleRegistry implements ModuleRegistryInterface {
  private readonly logger;
  private readonly moduleServices: ModuleService[] = [];
  constructor(private readonly logEvents: boolean, private readonly assetFactory = new AssetFactory()) {
    this.logger = new Logger("ModuleRegistry", logEvents);
  }

  add(config: ModuleConfig): ModuleService {
    const registeredAssets: string[] = this.moduleServices.reduce<string[]>(
      (prev, service) => service.assetSources.concat(prev),
      []
    );

    const intersection = registeredAssets.filter((value) =>
      config.assets.map((asset) => asset.src).includes(value)
    );

    if (intersection.length > 0) {
      throw Error(
        `A module including same sources: ${intersection} has already been registered. ${JSON.stringify(
          config
        )}`
      );
    }

    if (config.module) {
      const registeredModule: string[] = this.moduleServices
        .map((service) => service.module)
        .filter((item): item is string => !!item);

      if (registeredModule.includes(config.module)) {
        throw Error(
          `A module with same name: ${
            config.module
          } has already been registered. ${JSON.stringify(config)}`
        );
      }
    }

    return this.addModule(config);
  }

  // overrides existing module service if one is already registered
  addReplace(config: ModuleConfig): ModuleService {
    if (config.module) {
      let foundIndex = -1;
      const service = this.moduleServices.find((service, index) => {
        foundIndex = index;
        return service.module && service.module === config.module;
      });

      if (service && foundIndex > -1) {
        this.moduleServices.splice(foundIndex, 1);
        service.remove();
      }
    }

    return this.addModule(config);
  }

  private addModule(moduleConfig: ModuleConfig) {
    if (moduleConfig.type === "page") {
      const service = new PageModuleService(moduleConfig, this.assetFactory, this.logEvents);
      this.logger.log("registered page module: ", moduleConfig.module);
      this.moduleServices.push(service);
      return service;
    } else {
      const service = new ObservedModuleService(moduleConfig, this.assetFactory, this.logEvents);
      this.logger.log("registered observed module: ", moduleConfig.module);
      this.moduleServices.push(service);
      return service;
    }
  }
}

export default ModuleRegistry;
