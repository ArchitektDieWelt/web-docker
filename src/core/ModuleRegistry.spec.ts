import { describe, it, expect, afterEach, beforeEach } from "vitest";
import ModuleRegistry from "~/core/ModuleRegistry";
import { Asset } from "~/core/Asset";
import AssetFactory from "~/core/AssetFactory";
import { ModuleConfig } from "~/core/ModuleConfig";

const asset: Asset = {
  async: false,
  name: "chunk1.js",
  position: undefined,
  priority: 1,
  src: "http://src/1",
  deferred: false,
  type: "js",
};

const config: ModuleConfig = {
  pages: [],
  module: "test-module",
  type: "page",
  version: "1.0.0",
  assets: [asset],
};

afterEach(() => {
  document.body.innerHTML = "";
  document.head.innerHTML = "";
});

describe("ModuleRegistry", () => {
  it("constructs", () => {
    const moduleRegistry = new ModuleRegistry(false);

    expect(moduleRegistry).toBeTruthy();
  });
  it("adds new component service", async () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    const service = moduleRegistry.add(config);

    expect(service.module).toBe("test-module");
  });

  it("does not allow services with similar custom component names", () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    moduleRegistry.add(config);

    expect(() => moduleRegistry.add(config)).toThrow();
  });

  it("does not allow services with similar journey names", () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    moduleRegistry.add(config);

    expect(() => moduleRegistry.add(config)).toThrow();
  });

  it("does not allow services with duplicate assets", () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    const assetB: Asset = {
      async: false,
      name: "chunk2.js",
      position: undefined,
      priority: 1,
      src: "http://src/1",
      deferred: false,
      type: "js",
    };

    const configA: ModuleConfig = {
      pages: [],
      module: "test",
      type: "page",
      version: "1.0.0",
      assets: [asset, assetB],
    };

    moduleRegistry.add(config);

    expect(() => moduleRegistry.add(configA)).toThrow();
  });

  it("overrides service with similar module names", () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    const initialConfig: ModuleConfig = {
      pages: [],
      type: "page",
      version: "1.0.0",
      assets: [asset],
      module: "test-module",
    };

    moduleRegistry.add(initialConfig);

    const configOverride: ModuleConfig = {
      pages: [],
      version: "1.0.0",
      type: "page",
      assets: [asset],
      module: "test-module",
    };

    moduleRegistry.addReplace(configOverride);

    expect(document.body).toMatchSnapshot();
  });


  it("adds a shared page module", () => {
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const moduleRegistry = new ModuleRegistry(false, assetFactoryMock);

    const initialConfig: ModuleConfig = {
      pages: [],
      type: "page",
      version: "1.0.0",
      assets: [asset],
      module: "test-module",
      share: {
        name: "test",
        version: "1.0.0",
      }
    };

    moduleRegistry.add(initialConfig);

    expect(document.head).toMatchSnapshot();
  });
});
