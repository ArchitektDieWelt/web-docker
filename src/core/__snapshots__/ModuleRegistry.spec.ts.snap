// Vitest Snapshot v1

exports[`ModuleRegistry > adds new component service 1`] = `
<body>
  <link />
  <testcomponenta />
</body>
`;

exports[`ModuleRegistry > overrides service with similar journey names 1`] = `
<body>
  <link />
  <testcomponent />
</body>
`;

exports[`ModuleRegistry > overrides service with similar module names 1`] = `
<body>
  <link />
  <link />
</body>
`;
